### PR TITLE
apiserver/storage/cacher: consistent read from cache supports limit

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 1]
```diff
@@ -883,7 +910,19 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 	}
 	span.AddEvent("Filtered items", attribute.Int("count", listVal.Len()))
 	if c.versioner != nil {
-		if err := c.versioner.UpdateList(listObj, readResourceVersion, "", nil); err != nil {
+		var remainingItemCount *int64
+		var continueValue string
+		if applyLimit && len(objs) > len(selectedObjects) {
```
Comment (MadhavJivrajani) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:915]: Could you help me understand this? Can we not check `len(selectedObjects) > opts.Predicate.Limit`?

Also, `objs` comes from `c.ListItems`, which gets it from:
https://github.com/kubernetes/kubernetes/blob/b0263baae51895058b64b8e6f510a22a14ed8d3d/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L502

Here we try and get the first index that matches as a best effort scenario:
https://github.com/kubernetes/kubernetes/blob/b0263baae51895058b64b8e6f510a22a14ed8d3d/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L519-L528

and then do final filtering here:
https://github.com/kubernetes/kubernetes/blob/b0263baae51895058b64b8e6f510a22a14ed8d3d/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L869

So it might just be the case that `len(objs) == len(selectedObjects)`.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 2]
```diff
@@ -883,7 +910,19 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
...
+			continueValue, err = storage.EncodeContinue(string(lastSelectedObjectKey)+"\x00", key, int64(readResourceVersion))
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:926]: Serving pagination from cacher is out-of-scope for now.
For now, we want to serve only the first page from cacher.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 1 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: I think we should be fine since `WaitUntilFreshAndList` doesn't apply any filters, as the comment says:

```
// This isn't the place where we do "final filtering" - only some "prefiltering" is happening here. So the only
// requirement here is to NOT miss anything that should be returned. We can return as many non-matching items as we
// want - they will be filtered out later. The fact that we return less things is only further performance improvement.
```

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 3]
```diff
@@ -861,13 +890,21 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
...
 		elem, ok := obj.(*storeElement)
 		if !ok {
 			return fmt.Errorf("non *storeElement returned from storage: %v", obj)
 		}
 		if filter(elem.Key, elem.Labels, elem.Fields) {
 			selectedObjects = append(selectedObjects, elem.Object)
+			lastSelectedObjectKey = elem.Key
+		}
+		if applyLimit && int64(len(selectedObjects)) >= opts.Predicate.Limit {
+			hasMoreListItems = i < len(objs)-1
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:906]: alternatively we could loop over the remaining `objs` and see if there is at least one matching item.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 4]
```diff
@@ -754,14 +754,35 @@ func shouldDelegateList(opts storage.ListOptions) bool {
 	consistentReadFromStorage := resourceVersion == "" && !(consistentListFromCacheEnabled && requestWatchProgressSupported)
 	// Watch cache doesn't support continuations, so serve them from etcd.
 	hasContinuation := len(pred.Continue) > 0
-	// Serve paginated requests about revision "0" from watch cache to avoid overwhelming etcd.
-	hasLimit := pred.Limit > 0 && resourceVersion != "0"
+	// Serve paginated requests about revision "0" and "" (when consistent read from cache is supported)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:757]: Whenever you change this function, please also update the one in estimator - see the comment 10 lines above.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 3 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:898]: Let's not overoptimize.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 5]
```diff
@@ -221,7 +221,14 @@ func TestGetListCacheBypass(t *testing.T) {
 	})
 	t.Run("ConsistentListFromCache", func(t *testing.T) {
 		featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ConsistentListFromCache, true)
+		// TODO: fix me the following tests assume that etcdfeature.DefaultFeatureSupportChecker.Supports(storage.RequestWatchProgress)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:224]: We should address this.

I'm not sure I'm following this comment - are you saying that now we have a race that the feature is not necessarily initialized correctly when the test is running?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 6]
```diff
@@ -811,27 +811,11 @@ func (s *store) GetList(ctx context.Context, key string, opts storage.ListOption
...
-		var remainingItemCount *int64
-		// getResp.Count counts in objects that do not match the pred.
-		// Instead of returning inaccurate count for non-empty selectors, we return nil.
-		// Only set remainingItemCount if the predicate is empty.
-		if opts.Predicate.Empty() {
-			c := int64(getResp.Count - opts.Predicate.Limit)
-			remainingItemCount = &c
-		}
-		return s.versioner.UpdateList(listObj, uint64(withRev), next, remainingItemCount)
+	remainingItemCount, continueValue, err := storage.PrepareListRequestContinuationParametersIfApply(string(lastKey), keyPrefix, withRev, getResp.Count, hasMore, opts)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:814]: nit: can you please move just this refactor to a separate PR?

File: staging/src/k8s.io/apiserver/pkg/storage/util.go [hunk 7]
```diff
@@ -149,3 +149,30 @@ func HasInitialEventsEndBookmarkAnnotation(obj runtime.Object) (bool, error) {
 	objAnnotations := objMeta.GetAnnotations()
 	return objAnnotations[metav1.InitialEventsAnnotationKey] == "true", nil
 }
+
+// PrepareListRequestContinuationParametersIfApply prepares optional
+// parameters for retrieving additional results for a paginated request.
+//
+// This function sets up parameters that a client can use to fetch the remaining results
+// from the server if they are available.
+func PrepareListRequestContinuationParametersIfApply(keyLastItem, keyPrefix string, resourceVersion int64, itemsCount int64, hasMoreItems bool, opts ListOptions) (*int64, string, error) {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/util.go:158]: Let's just call it `PrepareContinue` (or `PrepareContinueToken`).

File: staging/src/k8s.io/apiserver/pkg/storage/util.go [hunk 8]
```diff
@@ -149,3 +149,30 @@ func HasInitialEventsEndBookmarkAnnotation(obj runtime.Object) (bool, error) {
 	objAnnotations := objMeta.GetAnnotations()
 	return objAnnotations[metav1.InitialEventsAnnotationKey] == "true", nil
 }
+
+// PrepareListRequestContinuationParametersIfApply prepares optional
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/util.go:153]: Let's put that in pkg/storage/continue.go

File: staging/src/k8s.io/apiserver/pkg/storage/util.go (hunk 7 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/util.go:158]: nit: and let's swap the order of result: `(string, *int64, error)`.
Also why *int64, not just int64 ?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go (hunk 5 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:224]: yes, I think so. The race exists today since this PR didn't introduce the test. This PR only adds a new case.

I have just run the `TestGetListCacheBypass` against the master branch and it fails:
```
=== RUN   TestGetListCacheBypass/ConsistentListFromCache
    cacher_whitebox_test.go:273: Unexpected bypass result for List request with options {ResourceVersion: ResourceVersionMatch: Predicate:{Label:<nil> Field:<nil> GetAttrs:<nil> IndexLabels:[] IndexFields:[] Limit:0 Continue: AllowWatchBookmarks:false} Recursive:false ProgressNotify:false SendInitialEvents:<nil>}, bypass expected: false, got: true
    --- FAIL: TestGetListCacheBypass/ConsistentListFromCache (0.00s)
 ```

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 6 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:814]: sure.

File: staging/src/k8s.io/apiserver/pkg/storage/util.go (hunk 7 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/util.go:158]: we need `*int64` because `remainingItemCount` will be used in (`nil` vs `0`)

https://github.com/kubernetes/kubernetes/blob/70e65eee46cbf72efabe8440a92117d1fedf6497/staging/src/k8s.io/apiserver/pkg/storage/interfaces.go#L55

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 4 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:757]: good catch, thanks.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:757]: do you know why we have two copies of the same function ?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go (hunk 5 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:224]: tmp fix in https://github.com/kubernetes/kubernetes/pull/125551

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 4 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:765]: Discussed offline.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 9]
```diff
@@ -762,14 +762,35 @@ func shouldDelegateList(opts storage.ListOptions) bool {
...
+	// from watch cache to avoid overwhelming etcd.
+	consistentListFromCache := resourceVersion == "" && (consistentListFromCacheEnabled && requestWatchProgressSupported)
+	hasLimit := pred.Limit > 0 && !(resourceVersion == "0" || consistentListFromCache)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:768]: Is this even needed at this point?

Given that you already implemented the support for limit in watchcache, it just works in watchcache. So existence of limit is no longer a reason to send request to etcd.
So can't we just remove the `hasLimit` completely?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 10]
```diff
@@ -762,14 +762,35 @@ func shouldDelegateList(opts storage.ListOptions) bool {
...
 
+// shouldApplyListLimit determines whether the cacher should
+// apply a limit to an incoming LIST request.
+//
+// note that this function doesn't check RVM nor the Continuation token.
+// these parameters are validated by the shouldDelegateList function.
+//
+// as of today, the limit is applied only to requests that
+// request a consistent snapshot of data from the cache.
+func shouldApplyListLimit(opts storage.ListOptions, requestWatchProgressSupported bool) bool {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:783]: Thinking more about this - instead of returning bool, can we instead just return the int (as limit to be used)?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 11]
```diff
@@ -762,14 +762,35 @@ func shouldDelegateList(opts storage.ListOptions) bool {
...
+	if opts.Predicate.Limit <= 0 {
+		return false
+	}
+	consistentListFromCacheEnabled := utilfeature.DefaultFeatureGate.Enabled(features.ConsistentListFromCache)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:787]: Following up on the comment about  removing `hasLimit` completely. Wouldn't this signficantly simplify this method too?

The only usecase we want to ignore the limit is the case of `resourceVersion='0' then, so this function would become trivial...

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 9 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:768]: I thought that we want to apply the limit only to requests that request a consistent snapshot of data from the cache as described in https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache#what-if-the-watch-cache-is-stale

If that is not the case then yeah we could get rid of `hasLimit` and always serve the first page from the cache.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:768]: That was my original intention, but thinking more about that, I think it doesn't make much sense.
In order to handle most of list requests (including those from kubectl) we need to support limit in watchcache.

For cases where we already handle a given semantic with limit unset in watchcache, as soon as we support the limit, it just doesn't make any sense to continue sending them to etcd.

So I'm treating this PR as somewhat orthogonal:
- implement "limit" in watchcache [the implementation is fairly trivial]
- given we have it, use it everywhere we have [to amplify the impact of ConsistentListsFromCache]

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:768]: okay, sgtm, i will update the pr, it should require much changes.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 11 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:787]: yes, removing `hasLimit` will make this function much simpler.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 10 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:783]: we could but i'm not sure if this would help because we need to ignore the limit for our legacy case.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 12]
```diff
@@ -201,9 +201,8 @@ func TestGetListCacheBypass(t *testing.T) {
 		{opts: storage.ListOptions{ResourceVersion: "0", Predicate: storage.SelectionPredicate{Continue: "a"}}, expectBypass: true},
 		{opts: storage.ListOptions{ResourceVersion: "1", Predicate: storage.SelectionPredicate{Continue: "a"}}, expectBypass: true},
 
-		{opts: storage.ListOptions{ResourceVersion: "", Predicate: storage.SelectionPredicate{Limit: 500}}, expectBypass: true},
 		{opts: storage.ListOptions{ResourceVersion: "0", Predicate: storage.SelectionPredicate{Limit: 500}}, expectBypass: false},
-		{opts: storage.ListOptions{ResourceVersion: "1", Predicate: storage.SelectionPredicate{Limit: 500}}, expectBypass: true},
+		{opts: storage.ListOptions{ResourceVersion: "1", Predicate: storage.SelectionPredicate{Limit: 500}}, expectBypass: false},
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:205]: this case doesn't require `features.ConsistentListFromCache` to be set. I think this is okay because depending on RVM it translates to `Exact` or `NotOlderThan` semantics

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width_test.go [hunk 13]
```diff
@@ -87,7 +87,7 @@ func TestWorkEstimator(t *testing.T) {
 				"events.foo.bar": 699,
 			},
 			maxSeats:             10,
-			initialSeatsExpected: 8,
+			initialSeatsExpected: 7,
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width_test.go:90]: I don't know why it changed. You have more experience with PnF so please double check it.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 14]
```diff
@@ -214,6 +213,7 @@ func TestGetListCacheBypass(t *testing.T) {
 		featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ConsistentListFromCache, false)
 		testCases := append(commonTestCases,
 			testCase{opts: storage.ListOptions{ResourceVersion: ""}, expectBypass: true},
+			testCase{opts: storage.ListOptions{ResourceVersion: "", Predicate: storage.SelectionPredicate{Limit: 500}}, expectBypass: true},
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:216]: this case still requires `features.ConsistentListFromCache` to be set.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 15]
```diff
@@ -762,12 +762,27 @@ func shouldDelegateList(opts storage.ListOptions) bool {
 	consistentReadFromStorage := resourceVersion == "" && !(consistentListFromCacheEnabled && requestWatchProgressSupported)
 	// Watch cache doesn't support continuations, so serve them from etcd.
 	hasContinuation := len(pred.Continue) > 0
-	// Serve paginated requests about revision "0" from watch cache to avoid overwhelming etcd.
-	hasLimit := pred.Limit > 0 && resourceVersion != "0"
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:766]: I think that we don't have to check if `storage.RequestWatchProgress)` is supported (NotOlderThan, Any semantics).

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width_test.go (hunk 13 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width_test.go:90]: This is fine - with the change this request is now assumed to be served from watchcache and the cost of it is proportional to total number of objects of that type (699).
Previously it was proportional to 2*limit (798).

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 16]
```diff
@@ -762,12 +762,27 @@ func shouldDelegateList(opts storage.ListOptions) bool {
...
+}
+
+// shouldApplyListLimit determines whether the cacher should
+// apply a limit to an incoming LIST request.
+//
+// note that this function doesn't check RVM nor the Continuation token.
+// these parameters are validated by the shouldDelegateList function.
+//
+// as of today, the limit is ignored for requests that set RV == 0
+func shouldApplyListLimit(opts storage.ListOptions) bool {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:778]: I still think that returning int would be better:
```
if opts.Predicate.Limit <= 0 || opts.ResourceVersion == "0" {
  return 0
}
return opts.Predicate.Limit
```

and then below you effectively change to:
```
limit := computeListLimit(opts)
...
if limit > 0 && len(selectedObjects) >= limit {
   ...
}
```

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 15 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:766]: agree

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go (hunk 12 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:205]: Exact will always be delegated (and bypassed) thanks to "unsupportedMatch" (cases below). But unset is handled as NotOlderThan so it's fine.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go (hunk 14 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:216]: agree

**Comment by p0lyn0mial** [comment]
`TestListWithConsistentListFromCache/test_List_with_limit_at_current_resource_version` is flaky, after looking at the test [definition](https://github.com/kubernetes/kubernetes/blob/c207f967ac3122d3a2a5ae4befdc5ada22347323/staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go#L811) I think the expected semantic is `Exact`.

 I don't think we can easily support this semantic in the cache so I would delegate such requests to etcd, wdyt ?

**Comment by wojtek-t** [comment]
> TestListWithConsistentListFromCache/test_List_with_limit_at_current_resource_version is flaky, after looking at the test [definition](https://github.com/kubernetes/kubernetes/blob/c207f967ac3122d3a2a5ae4befdc5ada22347323/staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go#L811) I think the expected semantic is Exact.

Nope - we should update the test to expect the same thing as for NotOlderThan.
The reason is that without a limit set, it was already behaving as `NotOlderThan`, so we're only unifying it.

**Comment by p0lyn0mial** [comment]
>>TestListWithConsistentListFromCache/test_List_with_limit_at_current_resource_version is flaky, after looking at the test [definition](https://github.com/kubernetes/kubernetes/blob/c207f967ac3122d3a2a5ae4befdc5ada22347323/staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go#L811) I think the expected semantic is Exact.

>Nope - we should update the test to expect the same thing as for NotOlderThan.
The reason is that without a limit set, it was already behaving as NotOlderThan, so we're only unifying it.

Based on https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list it looks like it would break backward compatibility, no ?

**Comment by wojtek-t** [comment]
> Based on https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list it looks like it would break backward compatibility, no ?

OK - it's documented explicitly. So yes, we should leave it as it was.
We can discuss with API approvers independently if we want to change that later.

But yes, we should do what you wrote above.

**Comment by aojea** [comment]
@wojtek-t @p0lyn0mial this seems related to this new failures in CI 

https://storage.googleapis.com/k8s-triage/index.html?test=API%20chunking

> Kubernetes e2e suite [It] [sig-api-machinery] Servers with support for API chunking should return chunks of results for list calls [Conformance]


/cc @liggitt

**Comment by liggitt** [comment]
that test failed before merge as well:

* https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125405/pull-kubernetes-e2e-gce/1800849655932653568
* https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125405/pull-kubernetes-e2e-ec2/1800849655815213056

please root cause and deflake or rollback

**Comment by Rajalakshmi-Girish** [comment]
https://prow.ppc64le-cloud.cis.ibm.net/job-history/gs/ppc64le-kubernetes/logs/periodic-kubernetes-containerd-conformance-test-ppc64le job also started failing after this PR.

**Comment by wojtek-t** [comment]
Same issue

**Comment by pacoxu** [comment]
> that test failed before merge as well:
> 
> * https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125405/pull-kubernetes-e2e-gce/1800849655932653568
> * https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125405/pull-kubernetes-e2e-ec2/1800849655815213056
> 
> please root cause and deflake or rollback

it add a new test in the package which is almost timeout for 180s.

See https://github.com/kubernetes/kubernetes/issues/125688

Some PR merge are blocked by the flake. I would suggest reverting this one temporarily.
